src/neo-vm/ExecutionEngine.cs

```diff
@@ -1154,6 +1192,123 @@ private bool ExecuteInstruction()
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal bool ExecuteEndTryCatch(TryState currentState)
+        {
+            var currentTry = CurrentContext.ErrorHandle?.CurContext;
+            if (currentTry is null) return false;
+            if (currentTry.State != currentState) return false;
+
+            if (currentTry.HasFinally)
+            {
+                currentTry.State = TryState.Finally;
+                currentTry.EndTryCatch(CurrentContext.InstructionPointer + CurrentContext.CurrentInstruction.Size);
+                CurrentContext.InstructionPointer = currentTry.FinallyPointer;
+            }
+            else
+            {
+                CurrentContext.ErrorHandle.Pop();
+                int nextOpcodePos = checked(CurrentContext.InstructionPointer + CurrentContext.CurrentInstruction.Size);
+                CurrentContext = InvocationStack.Peek();
+                CurrentContext.InstructionPointer = nextOpcodePos;
+            }
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ExecuteEndFinally()
+        {
+            var currentTry = CurrentContext.ErrorHandle?.Pop();
+            if (currentTry is null) return false;
+
+            if (FaultState.Rethrow)
+            {
+                FaultState.Rethrow = false;
+                throw FaultState.Error;
+            }
+            CurrentContext.InstructionPointer = currentTry.EndPointer;
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ExecuteThrow(string errorinfo)
+        {
+            var error = new CatcheableError(errorinfo);
+            var currentTry = CurrentContext.ErrorHandle?.CurContext;
+            if (currentTry != null)
+            {
+                if (currentTry.HasFinally)
+                {
+                    if (currentTry.State == TryState.Catch)
+                    {
+                        FaultState.Rethrow = true;
+                    }
+                    if (currentTry.State == TryState.Try && currentTry.HasCatch == false)
+                    {
+                        FaultState.Rethrow = true;
+                    }
+                }
+
+                if (currentTry.State == TryState.Try)
+                {
+                    if (currentTry.HasCatch == false && currentTry.HasFinally == false)
+                    {
+                        CurrentContext.ErrorHandle.Pop(); // if no final, no try, just pop try content.
+                    }
+                }
+                else if (currentTry.State == TryState.Catch)
+                {
+                    if (currentTry.HasFinally == false) // if has catch, but no final, just pop try content.
+                    {
+                        CurrentContext.ErrorHandle.Pop();
+                    }
+                }
+                else if (currentTry.State == TryState.Finally)
+                {
```

Comment:
@lightszero  Sine we have `FaultState.HandleError`, I think it's better to move this logic to there, just keep `ExecuteThrow` one line as:
```c#
throw new CatcheableError(errorinfo);
```